hal/src/nRF52840/ble_hal.cpp

```diff
@@ -15,739 +15,3147 @@
+    int setDeviceAddress(const hal_ble_addr_t* address);
+    int getDeviceAddress(hal_ble_addr_t* address);
+    int setAppearance(uint16_t appearance);
+    int getAppearance(uint16_t* appearance);
+    int addWhitelist(const hal_ble_addr_t* addrList, uint8_t len);
+    int deleteWhitelist(void);
 
-// GATT module instance
-NRF_BLE_GATT_DEF(g_gatt);
+private:
+    ble_gap_addr_t whitelist_[BLE_MAX_WHITELIST_ADDR_COUNT];                /**< BLE device whitelist. */
+    ble_gap_addr_t const* whitelistPointer_[BLE_MAX_WHITELIST_ADDR_COUNT];  /**< BLE device whitelists' pointer. */
+    BleObject* bleObj_;                                                     /**< Pointer to the BLE Object. */
 
-struct Char {
-    ble_uuid_t uuid;
-    ble_gatts_char_handles_t handles;
-    uint16_t flags;
+    static void processBleGapEvents(const ble_evt_t* event, void* context);
 };
 
-struct Service {
-    ble_uuid_t uuid;
-    uint16_t handle;
-    Char chars[BLE_MAX_CHAR_COUNT];
-    uint16_t charCount;
+class BleObject::Broadcaster {
+public:
+    Broadcaster(BleObject* obj);
+    ~Broadcaster() = default;
+    int init(void);
+    bool advertising(void) const;
+    int setAdvertisingParams(const hal_ble_adv_params_t* params);
+    int getAdvertisingParams(hal_ble_adv_params_t* params) const;
+    int setAdvertisingData(const uint8_t* buf, size_t len);
+    size_t getAdvertisingData(uint8_t* buf, size_t len) const;
+    int setScanResponseData(const uint8_t* buf, size_t len);
+    size_t getScanResponseData(uint8_t* buf, size_t len) const;
+    int setTxPower(int8_t val);
+    int8_t getTxPower(void) const;
+    int startAdvertising(void);
+    int stopAdvertising(void);
+
+private:
+    uint8_t isAdvertising_ : 1;                     /**< If it is advertising or not. */
+    uint8_t advHandle_;                             /**< Advertising handle. */
+    hal_ble_adv_params_t advParams_;                /**< Current advertising parameters. */
+    uint8_t advData_[BLE_MAX_ADV_DATA_LEN];         /**< Current advertising data. */
+    size_t  advDataLen_;                            /**< Current advertising data length. */
+    uint8_t scanRespData_[BLE_MAX_ADV_DATA_LEN];    /**< Current scan response data. */
+    size_t  scanRespDataLen_;                       /**< Current scan response data length. */
+    int8_t  txPower_;                               /**< TX Power. */
+    bool advPending_;                               /**< Advertising is pending. */
+    uint16_t connHandle_;                           /**< Connection handle. It is assigned once device is connected
+                                                         as Peripheral. It is used for re-start advertising. */
+    BleObject* bleObj_;                             /**< Pointer to the BLE Object. */
+    static const int8_t validTxPower_[8];           /**< Valid TX power values. */
+
+    int suspend(void);
+    int resume(void);
+    void toPlatformAdvData(ble_gap_adv_data_t* data);
```

Comment:
Consider returning by value.

Reply:
An `int` value for indicating error code?

Reply:
I meant `ble_gap_adv_data_t toPlatformAdvData()`